# Activity Data Visualization

This notebook contains the process followed to visualize activity.

## 1. Data

The data comes from a Garmin Forerunner 245 watch that has recorded activities for around 5 years. It can be downloaded using the Garmin Connect's dashpage into a CSV format.

### 1.1 Reading data.

The data is loaded from the `Activities.csv` file that should be located in the same directory as this notebook.

In [1]:
import pandas as pd

df = pd.read_csv("./Activities.csv")

df.head()

,Activity Type,Date,Favorite,Title,Distance,Calories,Time,Avg HR,Max HR,Aerobic TE,...,Steps,Total Reps,Total Sets,Decompression,Best Lap Time,Number of Laps,Moving Time,Elapsed Time,Min Elevation,Max Elevation
0,Walking,2026-03-19 20:04:39,False,Majadahonda Walking,6.41,472,01:06:17,120,150,3.1,...,"7,482",--,--,No,00:04:16.7,7,01:06:04,01:06:17,677,750
1,Walking,2026-03-18 10:46:40,False,Majadahonda Walking,1.64,103,00:17:40,124,142,1.0,...,"1,968",--,--,No,00:06:52.3,2,00:17:40,00:17:40,679,722
2,Trail Running,2026-03-18 10:13:00,False,Majadahonda - R20 W04 (02:00),3.48,268,00:33:08,140,166,2.6,...,"4,230",--,--,No,00:00:07.8,17,00:33:07,00:33:08,716,750
3,Walking,2026-03-18 10:02:07,False,Majadahonda Walking,1.01,82,00:10:08,138,158,2.0,...,"1,210",--,--,No,00:00:02.1,2,00:10:07,00:10:08,689,722
4,Walking,2026-03-17 10:59:00,False,Majadahonda Walking,1.01,65,00:11:00,126,134,0.7,...,"1,204",--,--,No,00:00:05.5,2,00:10:58,00:11:00,679,715


## 2. Understanding the data

By using the function `df.info()` we can see dataframe info, which showcases the following:

- 31 different columns.
- Default types set to `str`, `bool`, `float64`, `int64`.
- Some columns seem to have the wrong data type (`Distance` having `str` even most of its values are numeric).

| \# | Column                 | Non-Null Count | Dtype   |
|----|------------------------|----------------|---------|  
| 0  | Activity Type          | 180 non-null   | str     |
| 1  | Date                   | 180 non-null   | str     |
| 2  | Favorite               | 180 non-null   | bool    |
| 3  | Title                  | 180 non-null   | str     |
| 4  | Distance               | 180 non-null   | str     |
| 5  | Calories               | 180 non-null   | str     |
| 6  | Time                   | 180 non-null   | str     |
| 7  | Avg HR                 | 180 non-null   | str     |
| 8  | Max HR                 | 180 non-null   | str     |
| 9  | Aerobic TE             | 180 non-null   | str     |
| 10 | Avg Cadence            | 180 non-null   | str     |
| 11 | Max Cadence            | 180 non-null   | str     |
| 12 | Avg Pace               | 180 non-null   | str     |
| 13 | Best Pace              | 180 non-null   | str     |
| 14 | Total Ascent           | 180 non-null   | str     |
| 15 | Total Descent          | 180 non-null   | str     |
| 16 | Avg Stride Length      | 180 non-null   | str     |
| 17 | Training Stress Score® | 180 non-null   | float64 |
| 18 | Total Strokes          | 180 non-null   | str     |
| 19 | Avg. Swolf             | 180 non-null   | str     |
| 20 | Avg Stroke Rate        | 180 non-null   | str     |
| 21 | Steps                  | 180 non-null   | str     |
| 22 | Total Reps             | 180 non-null   | str     |
| 23 | Total Sets             | 180 non-null   | str     |
| 24 | Decompression          | 180 non-null   | str     |
| 25 | Best Lap Time          | 180 non-null   | str     |
| 26 | Number of Laps         | 180 non-null   | int64   |
| 27 | Moving Time            | 180 non-null   | str     |
| 28 | Elapsed Time           | 180 non-null   | str     |
| 29 | Min Elevation          | 180 non-null   | str     |
| 30 | Max Elevation          | 180 non-null   | str     |



In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Activity Type           180 non-null    str    
 1   Date                    180 non-null    str    
 2   Favorite                180 non-null    bool   
 3   Title                   180 non-null    str    
 4   Distance                180 non-null    str    
 5   Calories                180 non-null    str    
 6   Time                    180 non-null    str    
 7   Avg HR                  180 non-null    str    
 8   Max HR                  180 non-null    str    
 9   Aerobic TE              180 non-null    str    
 10  Avg Cadence             180 non-null    str    
 11  Max Cadence             180 non-null    str    
 12  Avg Pace                180 non-null    str    
 13  Best Pace               180 non-null    str    
 14  Total Ascent            180 non-null    str    
 15  

### 2.1 AI Analysis

I gave AI the following prompt to create me a table indicating it's suggested analysis of the data:

> Read the linked CSV file and analyze it, create a table with the following fields:
> 
> - Nº: Column number.
> - Column: Column name.
> - Description: Content description (what the data in the column usually looks like, excluding null indicators).
> - Null indicator: Indicator that seems to be used for when data is null.
> - Data type: Suggested data type to be used when reading CSV in Pandas.

*Model used: [qwen2.5:14b](https://ollama.com/library/qwen3.5)*

The AI then printed me the following table:

| Nº | Column               | Description                                                                                   | Null indicator | Data type         |
|----|----------------------|-----------------------------------------------------------------------------------------------|----------------|------------------|
| 1  | Activity Type        | Indicates the type of activity (e.g., Walking, Trail Running)                                  | --             | categorical       |
| 2  | Date                 | Timestamp indicating when the activity started                                                 | --             | datetime64[ns]    |
| 3  | Favorite             | Boolean value indicating if the activity is marked as a favorite                               | false          | boolean           |
| 4  | Title                | Name of the activity                                                                           | "Majadahonda Walking", etc. | string |
| 5  | Distance             | Total distance covered during the activity                                                      | --             | float64           |
| 6  | Calories             | Number of calories burned during the activity                                                   | --             | int64             |
| 7  | Time                 | Duration of the activity (hh:mm:ss)                                                            | --             | timedelta64      |
| 8  | Avg HR               | Average heart rate during the activity                                                          | --             | float64           |
| 9  | Max HR               | Maximum heart rate reached during the activity                                                  | --             | int64             |
| 10 | Aerobic TE           | Total energy expenditure from aerobic exercise                                                  | --             | float64           |
| 11 | Avg Cadence          | Average cadence (number of steps per minute)                                                    | --             | int64             |
| 12 | Max Cadence          | Maximum cadence reached during the activity                                                     | --             | int64             |
| 13 | Avg Pace            | Average pace of the activity in minutes per kilometer                                           | --             | timedelta64      |
| 14 | Best Pace           | Best pace achieved during the activity                                                           | --             | timedelta64      |
| 15 | Total Ascent         | Total ascent climbed during the activity                                                        | --             | float64           |
| 16 | Total Descent        | Total descent descended during the activity                                                     | --             | float64           |
| 17 | Avg Stride Length    | Average stride length during the activity                                                       | --             | float64           |
| 18 | Training Stress Score®| A score indicating overall stress on the body due to exercise                                 | --             | float64           |
| 19 | Total Strokes        | Number of strokes if applicable (swimming)                                                     | --             | int64             |
| 20 | Avg. Swolf           | Average swimming efficiency metric (synchronization of stroke and frequency)                    | --             | float64           |
| 21 | Avg Stroke Rate      | Average number of strokes per minute                                                            | --             | float64           |
| 22 | Steps                | Total number of steps taken during the activity                                                 | --             | int64             |
| 23 | Total Reps           | Total repetitions if applicable (strength training)                                            | --             | int64             |
| 24 | Total Sets           | Total sets if applicable (strength training)                                                    | --             | int64             |
| 25 | Decompression        | Indicates whether decompression was needed                                                      | No             | boolean           |
| 26 | Best Lap Time        | Best lap time achieved during the activity                                                     | "00:04:16.7", etc. | timedelta64      |
| 27 | Number of Laps       | Total number of laps completed                                                                 | --             | int64             |
| 28 | Moving Time          | The total moving time (excluding pauses)                                                        | "01:06:04"     | timedelta64      |
| 29 | Elapsed Time         | Total elapsed time including stops                                                             | "01:06:17"     | timedelta64      |
| 30 | Min Elevation        | Minimum elevation reached during the activity                                                   | --             | int64             |
| 31 | Max Elevation        | Maximum elevation reached during the activity                                                   | --             | int64             |

Although not perfect, this gives us a good starting point on cleaning up the data and having it be more readable. Some mistakes we can catch at first glanse from the analysis are the following:

- **Favorite:** `false` isn't necessarily a Null value, just means it hasn't been indicated.
- **Title:** Null indicator is clearly wrong (`"Majadahonda Walking"`is just a commonly used title).
- **Aerobic TE:** Didn't expect AI to get it right, it is [calculated based off training condition](https://www.garmin.com/en-US/garmin-technology/running-science/physiological-measurements/aerobic-training-effect/).
- **Decompression:** No doesn't necessarily mean its Null, just means it hasn't been indicated, could be changed to `True` and `False`.
- **Best Lap Time, Moving Time, Elapsed Time:** Null indicator clearly wrong, just a common time given.

Based off of the data analysis and browsing the CSV file

### 2.2 Cleaning data

Below are a couple steps taken to clean the data based off of the analysis the AI did.

In [3]:
# 1. Delete unused columns.
# These are columns which will not be used for further analysis and/or viz
try:
    df = df.drop(columns=[
        'Favorite', 'Total Reps', 'Total Sets', 'Decompression', 'Best Lap Time',
        'Avg Cadence', 'Total Strokes', 'Avg. Swolf', 'Elapsed Time', 
        'Number of Laps', 'Avg Stroke Rate', 'Avg Stride Length', 'Max Cadence',
        'Training Stress Score®'
        ])
except:
    print("ERROR: Couldn't drop columns, maybe you've already dropped them?")

# 2. Change Null indicator for None values when applicable.
# This will standardize values easier
for column in df.columns:
    df[column] = df[column].map(lambda x: None if x == '--' else x)

df.head()

# 3. Clean data
import re
import numpy as np

def clean_commas(x):
    if pd.isna(x):
        return np.nan
    return float(str(x).replace(",", ""))

# Remove commas from numbers (used to indicate thousanth place)
for col_name in ["Calories", "Steps", "Min Elevation", "Max Elevation"]:
    df[col_name] = df[col_name].apply(clean_commas)

# If data is not in correct format, switch it for None (for timedelta format)
def clean_timedelta(col: pd.Series, format="hhmmss") -> pd.Series:
    pattern = None
    if format == "mmss":  # MM:SS Format (Pace columns)
        pattern = re.compile(r"^\d{1,2}:\d{2}$")
    elif format == "hhmmss":  # HH:MM:SS (Time columns)
        pattern = re.compile(r"^\d{2}:\d{2}:\d{2}$")
    
    # keep only matching strings, else None
    clean_col =  col.map(lambda s: str(s) if pattern.match(str(s)) else None)
    return clean_col.astype("string")

df["Avg Pace"] = clean_timedelta(df["Avg Pace"], "mmss")
df["Best Pace"] = clean_timedelta(df["Best Pace"], "mmss")
df["Moving Time"] = clean_timedelta(df["Moving Time"], "hhmmss")
    
# 4. Change column type to correct one.
df["Date"] = pd.to_datetime(df["Date"])
df["Distance"] = pd.to_numeric(df["Distance"])
df["Calories"] = pd.to_numeric(df["Calories"]).astype("Int64")
df["Time"] = pd.to_timedelta(df["Time"])
df["Avg HR"] = pd.to_numeric(df["Avg HR"]).astype("Int64")
df["Max HR"] = pd.to_numeric(df["Max HR"]).astype("Int64") 
df["Aerobic TE"] = pd.to_numeric(df["Aerobic TE"])
df["Avg Pace"] = pd.to_timedelta("00:" + df["Avg Pace"].astype("string"), errors="coerce")
df["Best Pace"] = pd.to_timedelta("00:" + df["Best Pace"].astype("string"), errors="coerce")
df["Total Ascent"] = pd.to_numeric(df["Total Ascent"])
df["Total Descent"] = pd.to_numeric(df["Total Descent"])
df["Steps"] = pd.to_numeric(df["Steps"]).astype("Int64")
df["Moving Time"] = pd.to_timedelta(df["Moving Time"], errors="coerce")
df["Min Elevation"] = pd.to_numeric(df["Min Elevation"]).astype("Int64")
df["Max Elevation"] = pd.to_numeric(df["Max Elevation"]).astype("Int64")

# 5. Check new data types
df.info()
df.head()

<class 'pandas.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype          
---  ------         --------------  -----          
 0   Activity Type  180 non-null    str            
 1   Date           180 non-null    datetime64[us] 
 2   Title          180 non-null    str            
 3   Distance       179 non-null    float64        
 4   Calories       180 non-null    Int64          
 5   Time           180 non-null    timedelta64[us]
 6   Avg HR         178 non-null    Int64          
 7   Max HR         178 non-null    Int64          
 8   Aerobic TE     179 non-null    float64        
 9   Avg Pace       0 non-null      timedelta64[s] 
 10  Best Pace      0 non-null      timedelta64[s] 
 11  Total Ascent   159 non-null    float64        
 12  Total Descent  158 non-null    float64        
 13  Steps          175 non-null    Int64          
 14  Moving Time    165 non-null    timedelta64[us]
 15  Min Elevation  16

,Activity Type,Date,Title,Distance,Calories,Time,Avg HR,Max HR,Aerobic TE,Avg Pace,Best Pace,Total Ascent,Total Descent,Steps,Moving Time,Min Elevation,Max Elevation
0,Walking,2026-03-19 20:04:39,Majadahonda Walking,6.41,472,0 days 01:06:17,120,150,3.1,NaT,NaT,74.0,73.0,7482,0 days 01:06:04,677,750
1,Walking,2026-03-18 10:46:40,Majadahonda Walking,1.64,103,0 days 00:17:40,124,142,1.0,NaT,NaT,6.0,43.0,1968,0 days 00:17:40,679,722
2,Trail Running,2026-03-18 10:13:00,Majadahonda - R20 W04 (02:00),3.48,268,0 days 00:33:08,140,166,2.6,NaT,NaT,33.0,25.0,4230,0 days 00:33:07,716,750
3,Walking,2026-03-18 10:02:07,Majadahonda Walking,1.01,82,0 days 00:10:08,138,158,2.0,NaT,NaT,38.0,5.0,1210,0 days 00:10:07,689,722
4,Walking,2026-03-17 10:59:00,Majadahonda Walking,1.01,65,0 days 00:11:00,126,134,0.7,NaT,NaT,6.0,38.0,1204,0 days 00:10:58,679,715
